File: paicord-rs/PaicordLibBridge/Sources/DiscordHTTP/Types/Guild.swift
import PaicordLib

class BridgedGuild {
    var guild: Guild

    init (bridge: Guild) {
        self.guild = bridge
    }

    func get_id() -> UInt64 {
        return UInt64(self.guild.id.rawValue) ?? 0
    }

    func get_name() -> String {
        return self.guild.name
    }

    func get_icon() -> Optional<String> {
        return self.guild.icon
    }

    func get_banner() -> Optional<String> {
        return self.guild.banner
    }

    func get_channel_count() -> UInt {
        return UInt(self.guild.channels?.count ?? 0)
    }

    func get_channel(index: UInt) -> BridgedDiscordChannel {
        return BridgedDiscordChannel(channel: self.guild.channels![Int(index)])
    }

    func role_count() -> UInt {
        return UInt(self.guild.roles.count)
    }

    func get_role(index: UInt) -> RoleRust {
        return RoleRust(role: self.guild.roles[Int(index)])
    }
}

class BridgedPartialMember {
    var member: Guild.PartialMember

    init(member: Guild.PartialMember) {
        self.member = member
    }

    func has_user() -> Bool {
        return self.member.user != nil
    }

    func get_user() -> DiscordUserRust {
        return DiscordUserRust(user: self.member.user!)
    }

    func get_nick() -> Optional<String> {
        return self.member.nick
    }

    func get_avatar() -> Optional<String> {
        return self.member.avatar
    }

    func get_banner() -> Optional<String> {
        return self.member.banner
    }

    func get_pronouns() -> Optional<String> {
        return self.member.pronouns
    }

    func role_count() -> UInt {
        if let roles = self.member.roles {
            return UInt(roles.count)
        } else {
            return 0
        }
    }

    func get_role(index: UInt) -> SnowflakeRust {
        return SnowflakeRust(inner: UInt64(self.member.roles![Int(index)].rawValue) ?? 0)
    }

    func has_joined_at() -> Bool {
        return self.member.joined_at != nil
    }

    func get_joined_at() -> DiscordTimestampRust {
        return DiscordTimestampRust(timestamp: self.member.joined_at!)
    }

    func has_premium_since() -> Bool {
        return self.member.premium_since != nil
    }

    func get_premium_since() -> DiscordTimestampRust {
        return DiscordTimestampRust(timestamp: self.member.premium_since!)
    }

    func get_deaf() -> Bool {
        return self.member.deaf ?? false
    }

    func get_mute() -> Bool {
        return self.member.mute ?? false
    }

    func get_pending() -> Bool {
        return self.member.pending ?? false
    }

    func has_avatar_decoration() -> Bool {
        return self.member.avatar_decoration_data != nil
    }

    func get_avatar_decoration() -> AvatarDecorationRust {
        return AvatarDecorationRust(decoration: self.member.avatar_decoration_data!)
    }
}

File: paicord-rs/src/discord_models/types/guild.rs
use std::sync::Arc;

use crate::{
    discord_models::types::{channel::DiscordChannel, permission::Role, shared::DiscordTimestamp, snowflake::Snowflake, user::{AvatarDecoration, DiscordUser, PartialUser}},
    ffi,
};

#[derive(Clone, Debug)]
pub struct Guild {
    pub id: Snowflake,
    pub name: String,
    pub icon: Option<String>,
    pub banner: Option<String>,
    pub channels: Vec<DiscordChannel>,

    guild: Arc<ffi::BridgedGuild>,
}

impl std::fmt::Debug for ffi::BridgedGuild {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_tuple("BridgedGuild").finish()
    }
}

impl Guild {
    pub fn new(inner: ffi::BridgedGuild) -> Self {
        Self {
            id: Snowflake::from(inner.get_id()),
            name: inner.get_name(),
            icon: inner.get_icon(),
            banner: inner.get_banner(),
            channels: (0..inner.get_channel_count())
                .map(|i| inner.get_channel(i))
                .map(DiscordChannel::new)
                .collect(),
            guild: Arc::new(inner),
        }
    }

    pub fn get_roles(&self) -> Vec<Role> {
        (0..self.guild.role_count()).map(|i| self.guild.get_role(i)).collect()
    }

    pub fn get_members(&self) -> Vec<PartialMember> {
        (0..self.guild.member_count()).map(|i| self.guild.get_member(i)).map(PartialMember::new).collect()
    }
}

#[derive(Clone, Debug)]
pub struct PartialMember {
    pub user: Option<PartialUser>,
    pub nick: Option<String>,
    pub avatar: Option<String>,
    pub banner: Option<String>,
    pub pronouns: Option<String>,
    pub roles: Vec<Snowflake>,
    pub joined_at: Option<DiscordTimestamp>,
    pub premium_since: Option<DiscordTimestamp>,
    pub deaf: bool,
    pub mute: bool,
    pub pending: bool,
    pub avatar_decoration: Option<AvatarDecoration>,
}

impl PartialMember {
    pub fn new(inner: ffi::BridgedPartialMember) -> Self {
        Self {
            user: if inner.has_user() {
                Some(PartialUser::from(inner.get_user()))
            } else {
                None
            },
            nick: inner.get_nick(),
            avatar: inner.get_avatar(),
            banner: inner.get_banner(),
            pronouns: inner.get_pronouns(),
            roles: (0..inner.role_count()).map(|i| inner.get_role(i)).map(Snowflake::from).collect(),
            joined_at: if inner.has_joined_at() {
                Some(DiscordTimestamp::from(inner.get_joined_at()))
            } else {
                None
            },
            premium_since: if inner.has_premium_since() {
                Some(DiscordTimestamp::from(inner.get_premium_since()))
            } else {
                None
            },
            deaf: inner.get_deaf(),
            mute: inner.get_mute(),
            pending: inner.get_pending(),
            avatar_decoration: if inner.has_avatar_decoration() {
                Some(inner.get_avatar_decoration())
            } else {
                None
            },
        }
    }

    pub fn get_display_name(&self) -> Option<String> {
        if let Some(nick) = &self.nick {
            Some(nick.clone())
        } else if let Some(user) = &self.user {
            user.username.clone()
        } else {
            None
        }
    }
}

unsafe impl Send for ffi::BridgedGuild {}
unsafe impl Sync for ffi::BridgedGuild {}